examples/ex34.hpp

```diff
@@ -0,0 +1,819 @@
+//                                        MFEM Example 34
+
+#pragma once
+
+using namespace std;
+using namespace mfem;
+
+// Solver wrapper which orthogonalizes the input and output of part of a block vector
+// This is a modification of OrthoSolver in Navier miniapp
+class BlockOrthoSolver : public Solver
+{
+public:
+   BlockOrthoSolver(Array<int> &bOffsets_);
+
+   int global_size;
+
+   void SetSolver(Solver &s);
+
+   virtual void SetOperator(const Operator &op);
+
+   void Mult(const Vector &b, Vector &x) const;
+
+private:
+   Solver *solver = nullptr;
+
+   mutable Vector p_ortho, p_vec, v_vec;
+
+   mutable Vector temp;
+
+   Array<int> bOffsets, vblock, pblock;
+
+   void Orthogonalize(const Vector &v, Vector &v_ortho) const;
+};
+
+BlockOrthoSolver::BlockOrthoSolver(Array<int> &bOffsets_) : Solver(0, false),
+   bOffsets(bOffsets_)
+{
+   // form index array corresponding to velocity block
+   int start_ind = bOffsets[0];
+   int end_ind = bOffsets[1];
+   int vblock_size = bOffsets[1] - bOffsets[0];
+   vblock.SetSize(vblock_size);
+   for (int i=0; i < vblock_size; i++)
+   {
+      vblock[i] = start_ind + i;
+   }
+
+   // form index array corresponding to pressure block
+   start_ind = bOffsets[1];
+   end_ind = bOffsets[2];
+   int pblock_size = bOffsets[2] - bOffsets[1];
+   pblock.SetSize(pblock_size);
+   for (int i=0; i < pblock_size; i++)
+   {
+      pblock[i] = start_ind + i;
+   }
+}
+
+void BlockOrthoSolver::SetSolver(Solver &s)
+{
+   solver = &s;
+   height = s.Height();
+   width = s.Width();
+   MFEM_VERIFY(height == width, "Solver must be a square Operator!");
+}
+
+void BlockOrthoSolver::Mult(const Vector &b, Vector &x) const
+{
+   MFEM_VERIFY(solver, "Solver hasn't been set, call SetSolver() first.");
+   MFEM_VERIFY(height == solver->Height(),
+               "solver was modified externally! call SetSolver() again!");
+   MFEM_VERIFY(height == b.Size(), "incompatible input Vector size!");
+   MFEM_VERIFY(height == x.Size(), "incompatible output Vector size!");
+
+   // Orthogonalize input pressure block
+   b.GetSubVector(pblock,p_vec);
+   Orthogonalize(p_vec, p_ortho);
+   temp = b;
+   temp.SetSubVector(pblock,p_ortho); // update pressure block
+
+   // Propagate iterative_mode to the solver:
+   solver->iterative_mode = iterative_mode;
+
+   // Apply the Solver
+   solver->Mult(temp, x);
+
+   // Orthogonalize output
+   x.GetSubVector(pblock,p_vec);
+   Orthogonalize(p_vec,p_ortho);
+   temp = x;
+   temp.SetSubVector(pblock,p_ortho);
+}
+
+void BlockOrthoSolver::Orthogonalize(const Vector &v, Vector &v_ortho) const
```

Comment:
You can inherit from `OrthoSolver` and reuse the method (if other methods allow that). Or we should make the `OrthoSolver` one a static function.

Reply:
Yes that was my first try, but I ran into an issue with calling `Orthogonalize()` since its a private member function